pkg/daemon/controller/controller.go

```diff
@@ -203,6 +203,130 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
+}
+
+// removeServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the upgrade flag is false
+func (sc *ServiceController) removeServices(cm *core.ConfigMap, node core.Node, upgradeInPlace bool) error {
+	klog.Infof("removing services defined in ConfigMap %s", cm.Name)
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	for _, service := range cmData.Services {
+		if err := sc.ensureServiceIsRemoved(service.Name); err != nil {
+			return err
+		}
+	}
+
+	if !upgradeInPlace {
+		if err = sc.client.Delete(sc.ctx, &node); err != nil {
+			return errors.Wrapf(err, "error deleting node %s", node.GetName())
+		}
+		klog.Infof("deleted node %s", node.GetName())
+	}
+	return nil
+}
+
+// ensureServiceIsRemoved deletes the given service and all services that depend on it, even those unmanaged by WICD.
+func (sc *ServiceController) ensureServiceIsRemoved(service string) error {
+	existingSvcs, err := sc.getExistingServices()
+	if err != nil {
+		return errors.Wrap(err, "could not determine existing Windows services")
+	}
+	// Nothing to do if it already does not exist
+	if _, present := existingSvcs[service]; !present {
+		return nil
+	}
+
+	// Open the service to get a service handle
+	winSvcObj, err := sc.OpenService(service)
+	if err != nil {
+		return errors.Wrapf(err, "unable to open %s service handle", service)
+	}
+
+	status, err := winSvcObj.Query()
+	if err != nil {
+		return errors.Wrapf(err, "error querying %s service status", service)
+	}
+	// Delete the service and return if it's already stopped
+	if status.State == svc.Stopped {
+		if err := winSvcObj.Delete(); err != nil {
+			return errors.Wrapf(err, "failed to delete service %s", service)
+		}
+		klog.Infof("removed service %s", service)
+		return nil
+	}
+
+	// Ensure no dependent services are running before stopping this one. This is needed to safely stop Windows services
+	if err := sc.removeDependentServices(service, existingSvcs); err != nil {
+		return err
+	}
+
+	// Ensure service is stopped before deleting
+	if err := winsvc.EnsureServiceState(winSvcObj, svc.Stopped); err != nil {
+		return errors.Wrapf(err, "failed to stop service %s", service)
+	}
+	if err := winSvcObj.Delete(); err != nil {
+		return errors.Wrapf(err, "failed to delete service %s", service)
+	}
+	klog.Infof("removed service %s", service)
+	return nil
+}
+
+// removeDependentServices stops and deletes all services that depend on the given service
+func (sc *ServiceController) removeDependentServices(service string, existingSvcs map[string]struct{}) error {
+	for existingSvc := range existingSvcs {
+		existingWinSvcObj, err := sc.OpenService(existingSvc)
+		if err != nil {
+			return errors.Wrapf(err, "error opening %s service handle", existingSvc)
+		}
+		existingWinSvcConfig, err := existingWinSvcObj.Config()
+		if err != nil {
+			return errors.Wrapf(err, "error getting %s service config", existingSvc)
```

Comment:
Following the [above](https://github.com/openshift/windows-machine-config-operator/pull/1154/commits/93f50d7c64b18ea38f73d5c8dce98bd216c147cb#diff-e847d452a6040c690f8f6011bd0ecf612cc53aa12ce8ff997b29bb85870328f6R293) logging pattern, consider pushing the parameters to the end of the log message. For example:

```go
return errors.Wrapf(err, "error getting configuration for service %s", existingSvc)
```